### PR TITLE
Put Empty and Unit in Type0

### PR DESCRIPTION
**Pull request by SkySkimmer**
The only complicated fixing was in Modalities which needed changes in the universe annotations.
Universe minimization to Set is globally disabled so that we get lemmas like `Contr@{i} Unit` for all `i`. If Coq had cumulative inductive types like in http://advent-project.eu/pubs/cit-ictac15.pdf it might not be needed.
It is locally re-enabled in Idempotents to avoid a blowup in the number of universes (IsIdempotent goes from 4 to 51 universes), and for `Book_3_14` in the exercises to avoid some error.

### CODE REVIEW

**Comment by spitters** [comment]
LGTM, if Travis is happy.

**Comment by mikeshulman** [comment]
I'm confused.  I thought the point was to put Empty and Unit in the lowest universe, i.e. Set.  So why do they start acquiring universe annotations?

**Comment by mikeshulman** [comment]
Also I'm not real happy with hacky fixes when we don't understand what's going on.  Let's try to figure out what the problems are everywhere; if they are Coq bugs we should submit them, in any case we should at least understand what we're working around.

**Comment by SkySkimmer** [comment]
I made a less hacky fix for the exercise error. That one might be an underlying Coq bug.

Some of the new annotations are universes which used to be minimized to Set. For instance if you do
`Definition foo : 0 = 0 := idpath. About foo.` you get something like `foo@{Top.434} : @paths@{Top.434} nat O O`.

We need to disable minimization to get that kind of lemma though.
For instance, `Types.Unit.equiv_contr_unit` without minimization gets type `forall A : Type@{i}, Contr@{i} A -> Equiv@{i Set} A Unit`. However in `UnivalenceImpliesFunext.Univalence_implies_WeakFunext` we use it to prove `Equiv@{i i} (P x) Unit` for arbitrary `i`.
So it's either disable minimization or annotate equiv_contr_unit and a bunch of other lemmas.

Then since Empty/Unit aren't annotated any more they can't be used to determine which universe should be used so we get some more extra.
For instance under `A : Type@{i}`, `(@const A Unit@{i} tt)` becomes `@const@{i i} A Unit tt` in Modality.v, because that definition requires that universes not escape. Without annotations we get `const@{Top.13362 i}` and the new universe will stay forever.
Meanwhile definitions which weren't annotated just went through producing extra universes.

**Comment by JasonGross** [comment]
@mikeshulman:
> So why do they start acquiring universe annotations?

To give an much briefer answer: It's not `Unit` and `Empty` that acquire extra universe annotations, but everything which uses them in multiple places.

@SkySkimmer I'd like to see a timing diff for this change (`PREV_COMMIT=<sha1 of the version of master that you're basing this on> ./etc/coq-scripts/timing/make-pretty-timed-diff-tip.sh` and then look in `time-of-build-both.log`).  In particular, I want to see if the extra universes in things we don't annotate (e.g., `V.v`) make things slower.)

File: contrib/HoTTBookExercises.v [hunk 1]
```diff
@@ -517,7 +517,9 @@ Section Book_3_14.
       apply hprop_allpath.
       intros x' y'.
       etransitivity; [ symmetry; apply (p x x y' x') | ].
-      assert (H' : idpath = path_ishprop x x) by apply path_ishprop.
+     (* Without this it somehow proves [H'] using the wrong universe for hprop_Empty and fails when we do [Defined]. *)
```
Comment (JasonGross) [inline contrib/HoTTBookExercises.v]: This sounds like a Coq bug.  Would you be willing to run (on a version of HoTT without this change, but with all the other ones) [`/path/to/find-bug.py contrib/HoTTBookExercises.v bug_failed_defined_01.v -R theories HoTT --coqc ./hoqc --coqtop ./hoqtop`](https://github.com/JasonGross/coq-tools)?

**Comment by SkySkimmer** [comment]
@JasonGross 
>@SkySkimmer I'd like to see a timing diff for this change (PREV_COMMIT=<sha1 of the version of master that you're basing this on> ./etc/coq-scripts/timing/make-pretty-timed-diff-tip.sh and then look in time-of-build-both.log). In particular, I want to see if the extra universes in things we don't annotate (e.g., V.v) make things slower.)

In progress.

>This sounds like a Coq bug. Would you be willing to run (on a version of HoTT without this change, but with all the other ones) /path/to/find-bug.py contrib/HoTTBookExercises.v bug_failed_defined_01.v -R theories HoTT --coqc ./hoqc --coqtop ./hoqtop?

What is bug_failed_defined_01.v?

File: theories/Extensions.v [hunk 2]
```diff
@@ -326,7 +326,7 @@ Section Extensions.
   Definition ooExtendableAlong@{i j k l}
              {A : Type@{i}} {B : Type@{j}}
              (f : A -> B) (C : B -> Type@{k}) : Type@{l}
-    := forall n, ExtendableAlong@{i j k l} n f C.
+    := forall n : nat, ExtendableAlong@{i j k l} n f C.
```
Comment (JasonGross) [inline theories/Extensions.v]: Why is this annotation required?  Is it because otherwise we get a fresh universe `m` and an annotation `forall n : (nat : Type@{m}), ...`?  If that's the case, it seems like a bug in the setting (@mattam82); even when not minimizing most universes to `Set`, if a universe variable is not introduced by the user, and it's not passed to any identifier (only appears as a kind of phantom type annotation on a fully resolved type), then it's fine to minimize it to `Sed`.

File: theories/Idempotents.v [hunk 3]
```diff
@@ -428,7 +429,7 @@ Section Splitting.
 
   (** The splitting will be the sequential limit of the sequence [... -> X -> X -> X]. *)
   Definition split_idem : Type
-    := { a : nat -> X & forall n, f (a n.+1) = a n }.
+    := { a : nat -> X & forall n : nat, f (a n.+1) = a n }.
```
Comment (JasonGross) [inline theories/Idempotents.v]: This is required even given the `Local Set`ting at the top?

File: theories/Modalities/Accessible.v [hunk 4]
```diff
@@ -256,6 +256,7 @@ Module Accessible_Modalities_from_ReflectiveSubuniverses
     Definition inO_iff_isnull@{u a i} (O : Modality@{u a}) (X : Type@{i})
     : iff@{i i i} (In@{u a i} O X) (IsNull@{a i} (acc_gen O) X).
     Proof.
+      pose proof (@conn_map_to_O@{u a a a i a a a}).
```
Comment (JasonGross) [inline theories/Modalities/Accessible.v]: Why do we need this?

File: theories/Modalities/Identity.v [hunk 5]
```diff
@@ -84,13 +84,19 @@ Module Accessible_Identity <: Accessible_Modalities Identity_Modalities.
...
 
-  Definition inO_iff_isnull
+  Definition inO_iff_isnull@{u a i}
   : forall (O : Modality@{u a}) (X : Type@{i}),
       iff@{i i i}
         (In@{u a i} O X)
-        (IsNull@{u a i} (acc_gen O) X)
-    := fun O X => (fun _ => Empty_ind _ , fun _ => tt).
+        (IsNull@{u a i} (acc_gen@{u a} O) X).
+  Proof.
```
Comment (JasonGross) [inline theories/Modalities/Identity.v]: Did you say that you got an error if you gave this inline, or just that the universes were wrong?  Is there some set of universe annotations we can give (e.g., to `pair`) such that this works out nicely without the `Proof`?

**Comment by JasonGross** [comment]
> What is bug_failed_defined_01.v?

The name of the new file that the minimized test case will go into, once the script finishes running.

**Comment by SkySkimmer** [comment]
@JasonGross timing diff: https://gist.github.com/SkySkimmer/c989e2eb2c33f3c13945cab719fd47cd

**Comment by JasonGross** [comment]
Cool, thanks!

**Comment by SkySkimmer** [comment]
@JasonGross find-bug gets to this https://gist.github.com/SkySkimmer/be9fefed4a714c582d2bfd7d85c90ddc then it runs into changes in the error which I don't know how to adapt the regex to get past.

File: theories/Extensions.v (hunk 2 above)
Comment (SkySkimmer) [inline theories/Extensions.v]: > Is it because otherwise we get a fresh universe m and an annotation forall n : (nat : Type@{m}), ...?
Pretty much.

File: theories/Idempotents.v (hunk 3 above)
Comment (SkySkimmer) [inline theories/Idempotents.v]: No it isn't.

File: theories/Modalities/Accessible.v (hunk 4 above)
Comment (SkySkimmer) [inline theories/Modalities/Accessible.v]: Without it it makes a fresh universe instead of the `i` (twice because conn_map_to_O gets used twice).

**Comment by mikeshulman** [comment]
> It's not Unit and Empty that acquire extra universe annotations, but everything which uses them in multiple places.

That would make sense to me if it were true, but I actually see half a dozen places where Unit and Empty themselves have annotations.  That's what I don't understand.

File: theories/Extensions.v (hunk 2 above)
Comment (mikeshulman) [inline theories/Extensions.v]: Yikes!

**Comment by SkySkimmer** [comment]
@mikeshulman 
>That would make sense to me if it were true, but I actually see half a dozen places where Unit and Empty themselves have annotations. That's what I don't understand.

Could you specify a line? I'm not sure where you're looking.

File: theories/Extensions.v [hunk 6]
```diff
@@ -97,7 +97,7 @@ Section Extensions.
            (n : nat) {A : Type@{i}} {B : Type@{j}}
            (f : A -> B) (C : B -> Type@{k}) : Type@{l}
     := match n with
-         | 0 => Unit@{l}
```
Comment (mikeshulman) [inline theories/Extensions.v]: Here Unit has an annotation.

File: theories/Modalities/Accessible.v [hunk 7]
```diff
@@ -96,7 +96,7 @@ Record NullGenerators :=
 Coercion ngen_type : NullGenerators >-> Funclass.
 
 Definition null_to_local_generators : NullGenerators@{a1} -> LocalGenerators@{a2}
-  := fun S => Build_LocalGenerators (ngen_indices S) (ngen_type S) (fun _ => Unit@{a2}) (fun _ _ => tt).
```
Comment (mikeshulman) [inline theories/Modalities/Accessible.v]: Here Unit has an annotation.

File: theories/Modalities/Identity.v [hunk 8]
```diff
@@ -23,7 +23,7 @@ Module Identity_Modalities <: Modalities.
 
   Definition In : forall (O : Modality@{u a}),
                              Type@{i} -> Type@{i}
-    := fun O X => Unit@{i}.
```
Comment (mikeshulman) [inline theories/Modalities/Identity.v]: Here Unit has an annotation.  Etc.

File: theories/Extensions.v (hunk 6 above)
Comment (SkySkimmer) [inline theories/Extensions.v]: That's the current (master) version of the file.

**Comment by SkySkimmer** [comment]
Changed a couple things.

@mikeshulman All those are where I removed a type annotation on Unit/Empty.

**Comment by JasonGross** [comment]
@SkySkimmer When I run a fresh instance of `find-bug.py` on https://gist.github.com/SkySkimmer/be9fefed4a714c582d2bfd7d85c90ddc (telling it to input that file, and output into another file), it seems to not have much trouble?

**Comment by JasonGross** [comment]
@SkySkimmer reported as [bug #4862](https://coq.inria.fr/bugs/show_bug.cgi?id=4862)

File: theories/Modalities/Accessible.v [hunk 9]
```diff
@@ -125,7 +125,7 @@ Defined.
 
 Definition ooextendable_isnull_fibers {A B} (f : A -> B) (C : B -> Type)
 : (forall b, ooExtendableAlong (@const (hfiber f b) Unit tt)
-                               (fun _ => C b))
+                               (fun _ : Unit => C b))
```
Comment (mikeshulman) [inline theories/Modalities/Accessible.v]: Why is this necessary?

File: theories/Modalities/Notnot.v [hunk 10]
```diff
@@ -22,43 +22,42 @@ Module Notnot_Easy_Modalities <: EasyModalities.
...
     intros f z nBz.
     pose (unNotnot O).          (** Access the [Funext] hypothesis *)
-    (** The goal is [Empty@{j}], whereas [z] has codomain [Empty@{i}].  Thus, simply applying [z] would collapse the universe parameters undesirably, so we first alter the goal to be [Empty@{i}]. *)
-    cut (Empty@{i}); [ intros [] | ].
     apply z; intros a.
-    (** Now the goal is [Empty@{i}], whereas [f] has codomain [Empty@{j}]. *)
-    cut (Empty@{j}); [ intros [] | ].
-    exact (f a (transport (fun x => not@{j j j} (B x))
+    pose proof (hprop_Empty@{i}).
+    exact (f a (transport (fun x => not@{j j} (B x))
```
Comment (mikeshulman) [inline theories/Modalities/Notnot.v]: It's nice that we can avoid this hack now.

**Comment by mikeshulman** [comment]
`*`crawls into a hole`*` My brain must *really* have been turned off...

File: theories/Modalities/Identity.v [hunk 11]
```diff
@@ -84,13 +84,15 @@ Module Accessible_Identity <: Accessible_Modalities Identity_Modalities.
...
 
-  Definition inO_iff_isnull
+  Definition inO_iff_isnull@{u a i}
   : forall (O : Modality@{u a}) (X : Type@{i}),
       iff@{i i i}
         (In@{u a i} O X)
-        (IsNull@{u a i} (acc_gen O) X)
-    := fun O X => (fun _ => Empty_ind _ , fun _ => tt).
+        (IsNull@{u a i} (acc_gen@{u a} O) X)
+  := fun O X => @pair _ (_ -> Unit)
```
Comment (mikeshulman) [inline theories/Modalities/Identity.v]: What's going on here?

File: theories/Modalities/Open.v [hunk 12]
```diff
@@ -135,6 +135,8 @@ Module Accessible_OpenModalities <: Accessible_Modalities OpenModalities.
       refine (isequiv_ooextendable (fun _ => X) (@const (unOp O) Unit tt) ext).
   Defined.
 
+  Definition inO_iff_isnull@{u a i} := inO_iff_isnull'@{u a i i i i i i i i i i i i i i}.
+
```
Comment (mikeshulman) [inline theories/Modalities/Open.v]: It would be better if this hack could be avoided, but if not, then I'd like to at least have a comment in the file about what's going on.

**Comment by mikeshulman** [comment]
Ok, I'd like to know what's going on in `Idempotents` still, but aside from that I'm pretty happy.  We should include some remarks in `STYLE` about the sort of hacks that are required, like including extra type annotations `: nat` (unless that can be fixed) and `pose proof`ing lemmas so that typeclass inference doesn't find them more than once (it would be nice if typeclass inference could be made smarter about that too).

File: theories/Modalities/Accessible.v (hunk 9 above)
Comment (SkySkimmer) [inline theories/Modalities/Accessible.v]: Because it isn't.

File: theories/Modalities/Identity.v (hunk 11 above)
Comment (SkySkimmer) [inline theories/Modalities/Identity.v]: The new annotation on `acc_gen` isn't necessary, but it produces an unbound universe without the `(_ -> Unit)` bit.

File: theories/Modalities/Open.v (hunk 12 above)
Comment (SkySkimmer) [inline theories/Modalities/Open.v]: I reduced it some and made a comment, but couldn't avoid it completely.

File: contrib/HoTTBookExercises.v (hunk 1 above)
Comment (JasonGross) [inline contrib/HoTTBookExercises.v]: @SkySkimmer , could you add a reference to [bug #4862](https://coq.inria.fr/bugs/show_bug.cgi?id=4862) here?

File: theories/Types/Unit.v [hunk 13]
```diff
@@ -66,12 +66,12 @@ Global Instance isequiv_unit_rec `{Funext} (A : Type)
 (* For various reasons, it is typically more convenient to define functions out of the unit as constant maps, rather than [Unit_ind]. *)
 Notation unit_name x := (fun (_ : Unit) => x).
 
-Global Instance isequiv_unit_name `{Funext} (A : Type)
-: IsEquiv (fun (a:A) => unit_name a).
+Global Instance isequiv_unit_name `{Funext} (A : Type@{i})
```
Comment (JasonGross) [inline theories/Types/Unit.v]: I'm surprised you don't need `isequiv_unit_name@{i j}` here.  It's probably good practice to insert the `@{i j}` anyway.

**Comment by JasonGross** [comment]
Modulo adding a reference to the bug I mentioned in the comment, adding `@{i j}` to the instance, and adding a note to `STYLE` on adding explicit quantification of `nat` and `unit`, LGTM.

**Comment by JasonGross** [comment]
I reported the phantom binder issue as [bug #4868](https://coq.inria.fr/bugs/show_bug.cgi?id=4868).  I will add a mention of it to the style guide.

@mikeshulman after inspecting a few definitions, I think what's going on in modalities is that previously, there were a bunch of places where we had `Unit@{i}` and `Unit@{j}` in the same definition, and since `i` and `j` were not explicitly given by the user, Coq collapsed both to `Type1`.  Now, when it creates universes for `Unit` to live in, the universes are constrained only by `Set <= i` and `Set <= j`, and Coq does not identify `i` and `j`; see [bug #4870](https://coq.inria.fr/bugs/show_bug.cgi?id=4870).

I'm going to merge this, based on @spitters' and my acks.  Thanks @SkySkimmer !
